File: src/lib/serpapi/normalise.ts
import type { Engine } from "../config";
import type { BlockType, Evidence } from "../schemas";

interface RawResult {
  blockType: BlockType;
  title: string;
  url: string;
  domain: string;
  snippet: string;
  text?: string;
  date?: string;
  position: number;
  meta: Record<string, unknown>;
}

function domainOf(url: string): string {
  try {
    return new URL(url).hostname.replace(/^www\./, "");
  } catch {
    return "";
  }
}

function obj(v: unknown): Record<string, unknown> {
  return typeof v === "object" && v !== null ? (v as Record<string, unknown>) : {};
}

function arr(v: unknown): Record<string, unknown>[] {
  if (!Array.isArray(v)) return [];
  return v
    .filter((x): x is Record<string, unknown> => typeof x === "object" && x !== null)
    .map((x) => ({ ...x }));
}

function strOf(o: unknown, keys: string[]): string {
  const rec = obj(o);
  for (const k of keys) {
    const v = rec[k];
    if (typeof v === "string" && v.length > 0) return v;
  }
  return "";
}

function numOf(o: unknown, key: string): number {
  const v = obj(o)[key];
  const n = typeof v === "number" ? v : Number.parseFloat(String(v));
  return Number.isFinite(n) ? n : 0;
}

/** Normalise one engine's raw SerpApi JSON into Evidence rows. Every block is optional. */
export function normalise(engine: Engine, raw: Record<string, unknown>): RawResult[] {
  const out: RawResult[] = [];
  switch (engine) {
    case "google": {
      arr(raw.organic_results).forEach((o, i) => {
        const url = strOf(o, ["link", "website_link"]);
        out.push({
          blockType: "organic",
          title: strOf(o, ["title"]),
          url,
          domain: domainOf(url),
          snippet:
            strOf(o, ["snippet", "snippet_highlighted_words_snippet", "link"]) ||
            strOf(obj(o.rich_snippet).top, ["extensions"]),
          date:
            strOf(o, ["date"]) ||
            strOf(obj(o.publication_info).summary, ["summary", "source"]) ||
            strOf(obj(obj(o.about_this_result).source).description, ["description", "source"]),
          position: typeof o.position === "number" ? (o.position as number) : i + 1,
          meta: { sourceSpecific: o },
        });
      });
      arr(raw.related_questions).forEach((o, i) => {
        const url = strOf(o, ["link"]) || strOf(obj(arr(obj(o).snippet_links)[0]).link, ["link"]);
        const q = strOf(o, ["question"]);
        const snip = strOf(o, ["snippet"]);
        out.push({
          blockType: "related_question",
          title: q,
          url,
          domain: domainOf(url),
          snippet: q && snip ? `${q} — ${snip}` : (q || snip),
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      arr(raw.discussions_and_forums).forEach((o, i) => {
        const url = strOf(o, ["link"]);
        out.push({
          blockType: "forum",
          title: strOf(o, ["title"]),
          url,
          domain: domainOf(url),
          snippet: strOf(o, ["title"]) || strOf(obj(o.source).name, ["name"]),
          date: strOf(o, ["date"]),
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      arr(raw.related_searches).forEach((o, i) => {
        const q = strOf(o, ["query"]);
        out.push({
          blockType: "related_search",
          title: q,
          url: "",
          domain: "",
          snippet: q,
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      arr(raw.ads).forEach((o, i) => {
        const url = strOf(o, ["link", "displayed_link"]);
        out.push({
          blockType: "ad",
          title: strOf(o, ["title"]),
          url,
          domain: domainOf(url),
          snippet: strOf(o, ["snippet"]),
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      break;
    }
    case "google_news": {
      arr(raw.news_results).forEach((o, i) => {
        const url = strOf(o, ["link"]);
        const title = strOf(o, ["title"]);
        const sources = arr(o.sources);
        const pub = obj(o.publication_info);
        out.push({
          blockType: "news",
          title,
          url,
          domain: domainOf(url),
          snippet: strOf(o, ["snippet"]) || title,
          date:
            strOf(o, ["date"]) ||
            strOf(pub.published_time, ["basic"]) ||
            strOf(obj(sources[0]).title, ["title"]),
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      break;
    }
    case "google_autocomplete": {
      const s = arr(raw.suggestions);
      const list: unknown[] = s.length > 0 ? [...s] : (raw.suggestions as unknown[]);
      if (!Array.isArray(list)) break;
      list.forEach((v, i) => {
        const text = typeof v === "string" ? (v as string) : strOf(v, ["value"]);
        out.push({
          blockType: "suggestion",
          title: text,
          url: "",
          domain: "",
          snippet: text,
          position: i + 1,
          meta: {},
        });
      });
      break;
    }
    case "google_trends": {
      const data = arr(obj(raw.interest_over_time).timeline_data);
      data.forEach((point, i) => {
        const t = strOf(point, ["formattedTime", "time"]);
        const val = numOf(point, "value");
        out.push({
          blockType: "trend_point",
          title: `Interest ${t}`,
          url: "",
          domain: "trends",
          snippet: `${t}: ${val}`,
          text: String(val),
          position: i + 1,
          meta: { sourceSpecific: point },
        });
      });
      arr(raw.rising_related_queries).slice(0, 10).forEach((o, i) => {
        const q = strOf(o, ["query"]);
        out.push({
          blockType: "rising_query",
          title: q,
          url: "",
          domain: "trends",
          snippet: `Rising: ${q}`,
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      break;
    }
    case "google_shopping": {
      arr(raw.shopping_results).forEach((o, i) => {
        const url = strOf(o, ["link", "product_link"]);
        out.push({
          blockType: "product",
          title: strOf(o, ["title"]),
          url,
          domain: domainOf(url),
          snippet: [strOf(o, ["price"]), strOf(o, ["rating"]), strOf(o, ["reviews"]), strOf(o, ["source_name"])]
            .filter(Boolean)
            .join(" · "),
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      break;
    }
    case "google_play_product": {
      arr(raw.reviews).forEach((o, i) => {
        const url = strOf(o, ["link"]);
        out.push({
          blockType: "review",
          title: strOf(o, ["title", "user.name"]),
          url,
          domain: "play.google.com",
          snippet: strOf(o, ["snippet"]),
          date: strOf(o, ["date", "user.name"]),
          text: strOf(o, ["snippet"]),
          position: i + 1,
          meta: { rating: numOf(o, "rating"), sourceSpecific: o },
        });
      });
      break;
    }
    case "apple_reviews": {
      arr(raw.reviews).forEach((o, i) => {
        const url = strOf(o, ["link"]);
        out.push({
          blockType: "review",
          title: strOf(o, ["user.name"]),
          url,
          domain: "apps.apple.com",
          snippet: strOf(o, ["review"]),
          date: strOf(o, ["date", "user.name"]),
          text: strOf(o, ["review"]),
          position: i + 1,
          meta: { rating: numOf(o, "rating"), sourceSpecific: o },
        });
      });
      break;
    }
    case "google_maps": {
      arr(raw.local_results).forEach((o, i) => {
        out.push({
          blockType: "place",
          title: strOf(o, ["title"]),
          url: "",
          domain: strOf(o, ["domain"]),
          snippet: [strOf(o, ["rating"]), strOf(o, ["reviews"]), strOf(o, ["address"])]
            .filter(Boolean)
            .join(" · "),
          position: i + 1,
          meta: { rating: numOf(o, "rating"), dataId: strOf(o, ["data_id"]), sourceSpecific: o },
        });
      });
      break;
    }
    case "google_maps_reviews": {
      arr(raw.reviews).forEach((o, i) => {
        const snip = strOf(o, ["snippet", "text"]);
        out.push({
          blockType: "review",
          title: strOf(o, ["user.name"]),
          url: "",
          domain: "google.com/maps",
          snippet: snip,
          date: strOf(o, ["date"]),
          text: snip,
          position: i + 1,
          meta: { rating: numOf(o, "rating"), sourceSpecific: o },
        });
      });
      break;
    }
    case "google_jobs": {
      arr(raw.jobs_results).forEach((o, i) => {
        const rl = arr(obj(o).related_links);
        out.push({
          blockType: "job",
          title: strOf(o, ["title"]),
          url: strOf(rl[0], ["link"]),
          domain: "jobs",
          snippet: [strOf(o, ["company_name"]), strOf(o, ["location"]), strOf(o, ["description"])]
            .filter(Boolean)
            .join(" · "),
          date: strOf(obj(o).detected_extensions, ["posted_at"]),
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      break;
    }
    case "youtube": {
      arr(raw.video_results).forEach((o, i) => {
        const url = strOf(o, ["link"]);
        out.push({
          blockType: "organic",
          title: strOf(o, ["title"]),
          url,
          domain: domainOf(url),
          snippet: [strOf(obj(o).channel, ["name"]), strOf(o, ["length"]), strOf(o, ["published_time"])]
            .filter(Boolean)
            .join(" · "),
          date: strOf(o, ["published_time"]),
          position: i + 1,
          meta: { sourceSpecific: o },
        });
      });
      break;
    }
  }
  return out;
}

/** Wrap normalised rows into persisted Evidence entities for a run. */
export function toEvidence(
  runId: string,
  searchCallId: string,
  _engine: Engine,
  rows: RawResult[],
  startIndex: number,
): Evidence[] {
  return rows.map((r, i) => ({
    id: `E${startIndex + i + 1}`,
    runId,
    searchCallId,
    blockType: r.blockType,
    title: r.title,
    url: r.url,
    domain: r.domain,
    snippet: r.snippet,
    text: r.text,
    date: r.date,
    position: r.position,
    meta: r.meta,
  }));
}

export type { RawResult };

File: backend/app/normalise.py
"""Raw SerpApi JSON → evidence rows. Every block is optional; unknown shapes yield no rows, never an error."""
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from urllib.parse import urlparse


MAX_NEWS_ROWS = 15


def _obj(v: Any) -> Dict[str, Any]:
    return v if isinstance(v, dict) else {}


def _arr(v: Any) -> List[Dict[str, Any]]:
    return [x for x in v if isinstance(x, dict)] if isinstance(v, list) else []


def _str(o: Any, *keys: str) -> str:
    """First non-empty value; dotted keys walk nested objects ('user.name')."""
    for k in keys:
        v: Any = o
        for part in k.split("."):
            v = _obj(v).get(part)
        if isinstance(v, str) and v:
            return v
        if isinstance(v, (int, float)) and not isinstance(v, bool):
            return str(v)
    return ""


def _num(o: Any, key: str) -> float:
    try:
        return float(_obj(o).get(key))
    except (TypeError, ValueError):
        return 0.0


def _domain(url: str) -> str:
    try:
        host = urlparse(url).hostname or ""
    except ValueError:
        return ""
    return host[4:] if host.startswith("www.") else host


def _row(block: str, title: str, url: str, domain: str, snippet: str, position: int,
         date: Optional[str] = None, text: Optional[str] = None, meta: Optional[dict] = None) -> dict:
    row = {"blockType": block, "title": title, "url": url, "domain": domain, "snippet": snippet,
           "position": position, "meta": meta or {}}
    if date:
        row["date"] = date
    if text:
        row["text"] = text
    return row


def normalise(engine: str, raw: Dict[str, Any]) -> List[dict]:
    out: List[dict] = []
    if engine == "google":
        for i, o in enumerate(_arr(raw.get("organic_results"))):
            url = _str(o, "link")
            pos = o.get("position") if isinstance(o.get("position"), int) else i + 1
            out.append(_row("organic", _str(o, "title"), url, _domain(url), _str(o, "snippet") or _str(o, "title"), pos, _str(o, "date")))
        for i, o in enumerate(_arr(raw.get("related_questions"))):
            url, q, snip = _str(o, "link"), _str(o, "question"), _str(o, "snippet")
            out.append(_row("related_question", q, url, _domain(url), "%s — %s" % (q, snip) if q and snip else q or snip, i + 1))
        for i, o in enumerate(_arr(raw.get("discussions_and_forums"))):
            url = _str(o, "link")
            # the answers carry the actual user voice
            answers = " | ".join(a for a in (_str(x, "snippet") for x in _arr(o.get("answers"))[:3]) if a)
            out.append(_row("forum", _str(o, "title"), url, _domain(url), answers or _str(o, "title"), i + 1, _str(o, "date")))
        for i, o in enumerate(_arr(raw.get("related_searches"))):
            q = _str(o, "query")
            out.append(_row("related_search", q, "", "", q, i + 1))
        for i, o in enumerate(_arr(raw.get("ads"))):
            url = _str(o, "link", "displayed_link")
            out.append(_row("ad", _str(o, "title"), url, _domain(url), _str(o, "snippet"), i + 1))

    elif engine == "google_news":
        # a result is either one article or a cluster carrying `stories`
        items: List[dict] = []
        for o in _arr(raw.get("news_results")):
            items.extend(_arr(o.get("stories")) or [o])
        # Google News returns 80+ headlines per call; the top ones carry the signal and the
        # rest would drown user-voice rows (forums, People Also Ask) in the extractor
        for i, o in enumerate(items[:MAX_NEWS_ROWS]):
            url, title = _str(o, "link"), _str(o, "title")
            out.append(_row("news", title, url, _domain(url), _str(o, "snippet") or title, i + 1,
                            _str(o, "iso_date", "date"), meta={"source": _str(o, "source.name")}))

    elif engine == "google_autocomplete":
        for i, v in enumerate(raw.get("suggestions") if isinstance(raw.get("suggestions"), list) else []):
            text = v if isinstance(v, str) else _str(v, "value")
            if text:
                out.append(_row("suggestion", text, "", "", text, i + 1))

    elif engine == "google_trends":
        # TIMESERIES: timeline_data[] = {date, timestamp, values[{query, extracted_value}]}
        for i, point in enumerate(_arr(_obj(raw.get("interest_over_time")).get("timeline_data"))):
            label = _str(point, "date")
            values = {_str(v, "query").lower(): _num(v, "extracted_value") for v in _arr(point.get("values")) if _str(v, "query")}
            date = None
            try:
                date = datetime.fromtimestamp(int(_str(point, "timestamp")), tz=timezone.utc).isoformat()
            except ValueError:
                pass
            summary = ", ".join("%s: %g" % kv for kv in values.items())
            out.append(_row("trend_point", "Search interest %s" % label, "", "trends.google.com",
                            "%s — %s" % (label, summary), i + 1, date, meta={"values": values}))
        # RELATED_QUERIES: related_queries.rising[] = {query, value, extracted_value}
        for i, o in enumerate(_arr(_obj(raw.get("related_queries")).get("rising"))[:10]):
            q = _str(o, "query")
            out.append(_row("rising_query", q, _str(o, "link"), "trends.google.com",
                            "Rising search: %s (%s)" % (q, _str(o, "value") or "n/a"), i + 1,
                            meta={"value": _num(o, "extracted_value")}))

    elif engine == "google_shopping":
        for i, o in enumerate(_arr(raw.get("shopping_results"))):
            url = _str(o, "link", "product_link")
            bits = [_str(o, "price"), _str(o, "rating") and "%s★" % _str(o, "rating"),
                    _str(o, "reviews") and "%s reviews" % _str(o, "reviews"), _str(o, "source")]
            out.append(_row("product", _str(o, "title"), url, _domain(url), " · ".join(b for b in bits if b), i + 1,
                            meta={"price": _num(o, "extracted_price"), "rating": _num(o, "rating")}))

    elif engine == "google_play_product":
        for i, o in enumerate(_arr(raw.get("reviews"))):
            snip = _str(o, "snippet")
            out.append(_row("review", _str(o, "title", "user.name"), _str(o, "link"), "play.google.com", snip, i + 1,
                            _str(o, "iso_date", "date"), snip, {"rating": _num(o, "rating")}))

    elif engine == "apple_reviews":
        for i, o in enumerate(_arr(raw.get("reviews"))):
            snip = _str(o, "text", "review")
            out.append(_row("review", _str(o, "title", "user.name"), _str(o, "link"), "apps.apple.com", snip, i + 1,
                            _str(o, "review_date", "date"), snip, {"rating": _num(o, "rating")}))

    elif engine == "google_maps":
        for i, o in enumerate(_arr(raw.get("local_results"))):
            bits = [_str(o, "rating") and "%s★" % _str(o, "rating"), _str(o, "reviews") and "%s reviews" % _str(o, "reviews"),
                    _str(o, "type"), _str(o, "address")]
            out.append(_row("place", _str(o, "title"), "", _domain(_str(o, "website")) or "google.com/maps",
                            " · ".join(b for b in bits if b), i + 1, meta={"rating": _num(o, "rating"), "dataId": _str(o, "data_id")}))

    elif engine == "google_maps_reviews":
        for i, o in enumerate(_arr(raw.get("reviews"))):
            snip = _str(o, "snippet", "text")
            out.append(_row("review", _str(o, "user.name"), "", "google.com/maps", snip, i + 1,
                            _str(o, "iso_date", "date"), snip, {"rating": _num(o, "rating")}))

    elif engine == "google_jobs":
        for i, o in enumerate(_arr(raw.get("jobs_results"))):
            links = _arr(o.get("apply_options")) + _arr(o.get("related_links"))
            link = _str(links[0], "link") if links else ""
            bits = [_str(o, "company_name"), _str(o, "location"), _str(o, "description")[:300]]
            out.append(_row("job", _str(o, "title"), link or _str(o, "share_link"), _domain(link) or "google.com/jobs",
                            " · ".join(b for b in bits if b), i + 1, _str(o, "detected_extensions.posted_at")))

    elif engine == "youtube":
        for i, o in enumerate(_arr(raw.get("video_results"))):
            url = _str(o, "link")
            bits = [_str(o, "channel.name"), _str(o, "length"), _str(o, "published_date")]
            out.append(_row("organic", _str(o, "title"), url, _domain(url), " · ".join(b for b in bits if b), i + 1, _str(o, "published_date")))
    return out


def to_evidence(run_id: str, search_call_id: str, rows: List[dict], start_index: int) -> List[dict]:
    return [dict(r, id="E%d" % (start_index + i + 1), runId=run_id, searchCallId=search_call_id) for i, r in enumerate(rows)]
